docs/design/plan-modification.md

```diff
@@ -0,0 +1,543 @@
+
+
+### 3a. Composition and other helpers
+
+#### `All()`
+
+Similarly to `helper/schema`, the `All` composition helper runs all `RequiresReplaceFunc`s and returns true only if all funcs return true.
+
+```go
+func All(funcs ...RequiresReplaceFunc) bool {}
+```
+
+#### `Sequence()`
+
+Similarly to `helper/schema`, the `Sequence` composition helper runs all `RequiresReplaceFunc`s in sequence, stopping at the first that returns true. 
+
+```go
+func Sequence(funcs ...RequiresReplaceFunc) bool {}
+```
+
+#### `RequiresReplaceIfSet()`
+
+This helper reduces the verbosity of `RequiresReplaceFunc`s that always return false if either old or new value is null or unknown.
+
+
+This would require either:
+ - the addition of `IsNull()` and `IsUnknown()` functions to the `attr.Value` interface, or
+ - helper functions `attr.ValueIsKnown()` / `attr.ValueIsNull()` that use the `ToTerraformValue` to compare to `nil` and `tftypes.UnknownValue`
+ since there is at present no way to determine whether a generic `attr.Value` is null or unknown.
+
+Framework implementation:
+
+```go
+func RequiresReplaceIfSet(f RequiresReplaceFunc) RequiresReplaceFunc {
+  return func (ctx context.Context, old, new attr.Value) bool {
+    if old.IsNull() || old.IsUnknown() || new.IsNull() || new.IsUnknown() {
+      return false
+    }
+    return f(ctx, old, new)
+  }
+}
+```
+or
+```go
+func RequiresReplaceIfSet(f RequiresReplaceFunc) RequiresReplaceFunc {
+  return func (ctx context.Context, old, new attr.Value) bool {
+    if !attr.ValueIsKnown(old) || !attr.ValueIsKnown(new) || attr.ValueIsNull(old) || attr.ValueIsNull(new) {
+      return false
+    }
+    return f(ctx, old, new)
+  }
+}
+
+
+// in attr package
+
+func ValueIsKnown(ctx context.Context, val Value) (bool, error) {
+  v, err := val.ToTerraformValue(ctx)
+  if err != nil {
+    return false, err
+  }
+  return v == tftypes.UnknownValue
+}
+```
+
+Similar functions could be added for `attr.ValueIsFullyKnown()` and the like.
+
+### 4. `schema.Attribute.PlanModifiers`
+
+Extending the abstraction of `RequiresReplaceIf` one level higher, we can add a `PlanModifiers` field on the `schema.Attribute` struct, with the following framework code:
+
+```go
+type AttributePlanModifier interface {
+  Description(context.Context) string
+  MarkdownDescription(context.Context) string
+
+  Modify(context.Context, ModifyAttributePlanRequest, *ModifyAttributePlanResponse)
+}
+
+type AttributePlanModifiers []AttributePlanModifier
+
+type Attribute struct {
+  // ...
+  PlanModifiers AttributePlanModifiers
+}
+
+type ModifyAttributePlanRequest struct {
+	// ResourceConfig is the configuration the user supplied for the resource.
+	ResourceConfig Config
+
+	// ResourceState is the current state of the resource.
+	ResourceState State
+
+	// ResourcePlan is the planned new state for the resource.
+	ResourcePlan Plan
```

Comment:
You added these here, and then below say they're not available. I think because I made things confusing with the last review.

My hot take is let's omit them initially, but know that we can add them later, if we want, we're just waiting for a reason to. But if you feel they should be in from the get-go, I am fine with that answer to, so this is not a merge-blocking comment.

Reply:
Thanks, I meant to remove the bit that said they're not available. Let's see what @bflad thinks about having them in the initial version.

Reply:
Naming aside (would data sources have separate fields?), since this is just a design document my recommendation would be to fix the verbiage about availability. I think abandoned ideas or future ideas do have some importance to capture since those can be representative of everything the team was thinking about and aware of at the time of design -- whether and when those designs become reality is more of an implementation and prioritization question.